haystack/core/pipeline/component_checks.py

```diff
@@ -0,0 +1,244 @@
+
+    A component needs to pass two gates so that it is ready to run:
+    1. It has received all mandatory inputs.
+    2. It has received a trigger.
+    :param component: Component metadata and the component instance.
+    :param inputs: Inputs for the component.
+    """
+    received_all_mandatory_inputs = are_all_sockets_ready(component, inputs, only_check_mandatory=True)
+    received_trigger = has_any_trigger(component, inputs)
+
+    return received_all_mandatory_inputs and received_trigger
+
+
+def has_any_trigger(component: Dict, inputs: Dict) -> bool:
+    """
+    Checks if a component was triggered to execute.
+
+    There are 3 triggers:
+    1. A predecessor provided input to the component.
+    2. Input to the component was provided from outside the pipeline (e.g. user input).
+    3. The component does not receive input from any other components in the pipeline and `Pipeline.run` was called.
+
+    A trigger can only cause a component to execute ONCE because:
+    1. Components consume inputs from predecessors before execution (they are deleted).
+    2. Inputs from outside the pipeline can only trigger a component when it is executed for the first time.
+    3.  `Pipeline.run` can only trigger a component when it is executed for the first time.
+
+    :param component: Component metadata and the component instance.
+    :param inputs: Inputs for the component.
+    """
+    trigger_from_predecessor = any_predecessors_provided_input(component, inputs)
+    trigger_from_user = has_user_input(inputs) and component["visits"] == 0
+    trigger_without_inputs = can_not_receive_inputs_from_pipeline(component) and component["visits"] == 0
+
+    return trigger_from_predecessor or trigger_from_user or trigger_without_inputs
+
+
+def are_all_sockets_ready(component: Dict, inputs: Dict, only_check_mandatory: bool = False) -> bool:
+    """
+    Checks if all sockets of a component have enough inputs for the component to execute.
+
+    :param component: Component metadata and the component instance.
+    :param inputs: Inputs for the component.
+    :param only_check_mandatory: If only mandatory sockets should be checked.
+    """
+    filled_sockets = set()
+    expected_sockets = set()
+    if only_check_mandatory:
+        sockets_to_check = {
+            socket_name: socket for socket_name, socket in component["input_sockets"].items() if socket.is_mandatory
+        }
+    else:
+        sockets_to_check = {
+            socket_name: socket
+            for socket_name, socket in component["input_sockets"].items()
+            if socket.is_mandatory or len(socket.senders)
+        }
+
+    for socket_name, socket in sockets_to_check.items():
+        socket_inputs = inputs.get(socket_name, [])
+        expected_sockets.add(socket_name)
+        if (
+            is_socket_lazy_variadic(socket)
+            and any_socket_input_received(socket_inputs)
+            or has_socket_received_all_inputs(socket, socket_inputs)
+        ):
```

Comment:
A minor suggestion for code efficiency, and this might make it easier to understand the intended logic:
```suggestion
        # Check if socket has all required inputs or is a lazy variadic socket with any input

       if (
    has_socket_received_all_inputs(socket, socket_inputs)
    or (is_socket_lazy_variadic(socket) and any_socket_input_received(socket_inputs))
):
```